### ISSUE TITLE
Provide "enterprise-friendly" deployment

### ISSUE THREAD

**Issue by pascalberger** [ForkIssues/TrackerWin#57]
It would be nice to have a distribution of Fork which is more suitable for controlled environments (e.g. in enterprise environments). This affects installer, but also settings, which currently don't work really well in these kind of environments.

- [ ] Install system wide instead of per use, e.g. through using an MSI. This would also work better with system wide package managers like Chocolatey. See #10
- [x] ~~Use system Git instead of using internal Git version. This gives consistency through the whole tool chain can be important to make Fork working with a specific environment where a specific version of Git is required.~~ 
  Available since Git 1.40.
- [ ] Disable auto update, so that installation can be managed through software management tool (like Chocolatey)
- [ ] The possibility to configure Fork before first use.
- [ ] Non personal multi-user licensing, possibility to automatically deploy license

**Comment by claudenobs** [comment]
yeah, having an user-local installer is supernice, but an installer.msi that can use system git would be superdupernice.
for instance, at our company we use a custom git-flow version, so while we currently can use fork, we can't use it for the git-flow related workflows :-(

**Comment by JackZylkin** [comment]
Yes please allow us to use system-level git. My users do not know, and should not need to know, that the git-config commands they execute on the command line have no effect on Fork.

**Comment by DanPristupov** [comment]
> that the git-config commands they execute on the command line have no effect on Fork.

What do you mean?
`--global` values are stored in `%userprofile%/.gitconfig` and do make effect on Fork.

**Comment by javs1979** [comment]
Upping the current issue... for Windows version, whether MSI is used or not, for me, "enterprise-friendly" deployment is...

- [x] Ability to install Fork silently (without human intervention)
- [x] Ability to return valid or useful return code (`0` for success, use [Windows system codes for the rest](https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes))
- [x] Ability to provide install arguments (with or without human intervention)

As for install arguments...
- `/UserName` with current user name (derived from `git config user.name` if available, or `[System.DirectoryServices.AccountManagement.UserPrincipal]::Current.DisplayName;`
- `/Email` with current user email (derived from `git config user.email` if available, or `[System.DirectoryServices.AccountManagement.UserPrincipal]::Current.EmailAddress;`
- `/DefautCloneRepo` with given value (or any valid derived value from `git config` if available... or elsewhere)

What do you think ?

**Comment by DanPristupov** [comment]
There's no way I can get arguments from the Squirrel installer.

**Comment by javs1979** [comment]
> There's no way I can get arguments from the Squirrel installer.

Uh... I was thinking about the ability to **set** arguments **to** the Squirrel installer, so that installation runs silently.

Whether Fork installer is able to provide **default and appropriate values** to the arguments is not _that_ relevant, as long as the installer is able to read values (user name, email and default clone repo) from CLI and use them to run installation silently.

**Comment by pascalberger** [comment]
@javs1979 You already can run the installer silent using the `/s` parameter. Configuration on first startup is something different though

**Comment by javs1979** [comment]
> @javs1979 You already can run the installer silent using the `/s` parameter. Configuration on first startup is something different though

Well, that's a start... but are there any usable parameters alongside the `/s` switch ?

**Comment by sreece-pas** [comment]
Agree with @javs1979. Even if the installer can't pull the variables, just being able to set them via command line would mean I could use PDQ Deploy to push to users with the first step being pulling the values myself.

**Comment by DanPristupov** [comment]
What is the difference between Scoop and Chocolatey? Which one should we support?

**Comment by jobec** [comment]
Adding to the above list of @javs1979 

* The possibility to configure Fork before first use.

I now need to copy over a settings file and modify it according to the user's environment.
It would be easier to have, for example CLI arguments to `Fork.exe` to set certain settings.

**Comment by javs1979** [comment]
> 
> 
> Adding to the above list of @javs1979
> 
>     * The possibility to configure Fork before first use.
> 
> 
> I now need to copy over a settings file and modify it according to the user's environment.
> It would be easier to have, for example CLI arguments to `Fork.exe` to set certain settings.

Yes. That, added to the 'CLI-settable arguments' (as a usable alternative) would be very useful.

Reminds me of 'answer files' in Task sequence steps in SCCM, but with Fork, it'd be much simpler because of the minimum required arguments for any `git` instance (see [here](https://github.com/ForkIssues/TrackerWin/issues/57#issuecomment-542307853) for an example).

**Comment by pascalberger** [comment]
> What is the difference between Scoop and Chocolatey? Which one should we support?

Scoop (as recently announced WinGet) has a much more narrow scope and uses a repo as backend to install applications and is basically a wrapper around an installer while Chocolatey is a fully fledged package manager allowing more advanced scenarios. Regarding permissions Scoop installs to user directory, while Chocolatey is a system wide package manager. See also https://github.com/lukesampson/scoop/wiki/Chocolatey-Comparison. While both could used for something like Fork, Chocolatey would be a more natural fit (e.g. default Scoop repo doesn't contain GUI tools, scoop-extras does). Regarding this issue Chocolatey is definitely the tool to use in Enterprise scenarios.

Another point to add while Fork can't be used in Enterprise scenarios: People often don't have administration permissions. While not using admin permissions at all seems like a solution at first glance it won't work in a lot of cases, since there are often existing tools in place to handle this scenario, e.g. using Chocolatey Agent to allow installing approved Software under an account which has administrative permissions. Having a user level installer won't work with these kind of solutions.

**Comment by claudenobs** [comment]
scoop is awesome for cli tools, especially if they need to be on the PATH.
it's also great for portable (gui)-tools, that store their config in the application directory.

while it can install apps globally and works with tools that use installers, auto-update themself, need registry access, it's not made with them in mind.

generally i use choco for everything that needs admin permissions and/or benefits from systemwide/installer installation and use scoop for everything else (like rust, go, jdk, node, python, processhacker, keepass, freecad, inkscape, libreoffice, ...)

**Comment by github-kh2018** [comment]
I would also like to vote for machine-wide installation to get rid of the disease that is "installation to LOCALAPPDATA". Application white-listing with Applocker gets more complicated if you have to create exceptions for user-owned folders. Applications should be located in "ProgramFiles"!

**Comment by DanPristupov** [comment]
> Applications should be located in "ProgramFiles"

Well, this makes the automatic update process much more complicated (and less secure) because update must be performed by a separate service with admin rights (this is what Chromium does).

**Comment by pascalberger** [comment]
> Well, this makes the automatic update process much more complicated (and less secure) because update must be performed by a separate service with admin rights (this is what Chromium does).

Which is exactly what some enterprises require. Since the requirements for a deployment are very different and even contrary, I would suggest to go to still provide the squirrel based installer, but additionally provide an enterprise distribution which uses an MSI installer and has auto-update disabled. SourceTree for example has decided to go the same way after a very similar discussion and introduced SourceTree entrprise (see (https://www.sourcetreeapp.com/enterprise), an MSI based installer, without auto-update and without its own Git instance.

**Comment by pascalberger** [comment]
Added an additional point for non-personal multi-user licensing and possibility to automatically deploy license to initial list.

**Comment by github-kh2018** [comment]
> Well, this makes the automatic update process much more complicated (and less secure) because update must be performed by a separate service with admin rights (this is what Chromium does).

You are correct, a separate update service like the Mozilla maintenance service is more complex. But for the use case of enterprise deployment this would probably get disabled anyway. As a first step, I would suggest "installation to profile with automatic updates" XOR "installation to Program Files, NO automatic updates"

**Comment by DanPristupov** [comment]
I think we can implement the support for default config file, which will be used to configure Fork on first run. With the combination with the `/s` option, we can create an enterprise installer as a .bat file. This may help to automate the installation, however the system-wide installation is still not possible. 

Example:

forkinstall.bat:
``` 
@ECHO OFF
 
:: Run Fork installer silently
ForkInstaller.exe /s
 
:: Prepare the Fork settings file which will be used on first run
(
echo "user.name=username"
echo "user.email=email@company.com"
echo "defaultSourceDirectory=C:\sources"
echo "git=C:\program files\git\bin\git.exe"
echo "license=xxxxxx-xxxxxx-xxxxxx-xxxxxx-xxxxxx"
) > %localappdata%/Fork/setup.config
```

When you run it, it will:
1. Silently install Fork to `%localappdata%/Fork`
2. Create the `setup.config` file which will be read by Fork on first run.
Both steps will be performed without showing an UI.
 
When the user run Fork the first time, the application will configure the environment based on the settings file:
- extract the git instance and switch the git if needed
- activate the license
- setup the user identity which will be used to commit
- scan the repositories in the default source directory
 
What do you think?

**Comment by toogeneric** [comment]
Sounds fantastic to me!

For our company that'd be all we need.

**Comment by github-kh2018** [comment]
> 
> I think we can implement the support for default config file, which will be used to configure Fork on first run. With the combination with the `/s` option, we can create an enterprise installer as a .bat file. This may help to automate the installation, however the system-wide installation is still not possible.
>
I can often work around some "features" of an application and provide configuration for users or even try to disable auto-updates by brute force (ie. by host firewall rule) but if a program is not designed to install and run system-wide, that's it.
These days even Microsoft provides an MSI installer for MS Teams (although I have to make the installer believe it runs on a terminal server :-/).
So while I appreciate this step, the system-wide configuration is still the big issue.

**Comment by pascalberger** [comment]
While this solves some issues, others can't be solved this way and providing an MSI installer is probably the easiest option. E.g. if a system wide package manager is used a local installer leads to issues.

**Comment by jamaya77** [comment]
Hi,
Just in case someone is still looking for a solution, here's my two cents. One thing I'd like to note is that, while I agree that an MSI package for any application works better for deployment, it will not make any difference for Fork even if you had an MSI installation package. The problem I noticed with both ForkInstaller.exe and the converted MSI package is that the Trigger Action from the SCCM client, would fail to run and report an error after the installation. I don know why, but that was my observation even though Fork was installing correctly from Software Center.

Note: This application must be deployed while the user is logged-in, so making the application available via Software Center for self-deployment is the best approach.

So Here's my solution:
1) Create a `install.cmd` file to run the  commands in the following order:
```
@echo off

ForkInstaller.exe -s

echo %username%

timeout /T 10 /NOBREAK

%localappdata%\fork\fork.exe activate %username%@email.com ########-########-########-########-########

CScript machinepolicy.vbs
```

2) Create a `uninstall.cmd` file with the following commands in the same order:
```
@echo off

"%LocalAppData%\Fork\Update.exe" --uninstall -s

CScript machinepolicy.vbs
```

3) Create a vbs script (for this example call it machinepolicy.vbs) so that it runs the trigger action manually after the installation without running into permission issues:
```
'Declare Variables
On Error Resume Next
Set sho = CreateObject("WScript.Shell")
strSystemRoot = sho.expandenvironmentstrings("%SystemRoot%")
strCurrentDir = Left(Wscript.ScriptFullName, (InstrRev(Wscript.ScriptFullName, "\") -1))
' Get a connection to the "root\ccm\invagt" namespace (where the Inventory agent lives)
Dim oLocator
Set oLocator = CreateObject("WbemScripting.SWbemLocator")
Dim oServices
Set oServices = oLocator.ConnectServer( , "root\ccm\invagt")
'Reset SMS Hardware Inventory Action to force a full HW Inventory Action
sInventoryActionID = "{00000000-0000-0000-0000-000000000021}"
' Delete the specified InventoryActionStatus instance
oServices.Delete "InventoryActionStatus.InventoryActionID=""" & sInventoryActionID & """"
'Pause 3 seconds To allow the action to complete.
wscript.sleep 3000
'Run a SMS Hardware Inventory
Set cpApplet = CreateObject("CPAPPLET.CPAppletMgr")
Set actions = cpApplet.GetClientActions
For Each action In actions
If Instr(action.Name,"Hardware Inventory") > 0 Then
action.PerformAction
End If
Next
```

4) In SCCM, configure the "Installation Program" to run the `install.cmd` file:
`cmd.exe /c install.cmd /VERYSILENT /NORESTART`

5) Then configure the "Uninstall program" with the `uninstall.cmd` file:
`cmd.exe /c uninstall.cmd /VERYSILENT /NORESTART`

6) Configure the "Detection Method" with a powershell script:
```
###Fork###

if (Test-Path $("${Env:LOCALAPPDATA}"+"\Fork\Fork.exe"))
{
Write-Host "installed"
}
 
else
{
}
```

7) Set `Install Behavior` to `Install For User`.

**Comment by Lonli-Lokli** [comment]
My +1 for enterprise support.

For me the main problem is an auto-update feature, usually, it is not allowed.

**Comment by Lonli-Lokli** [comment]
@DanPristupov @jamaya77   Do you know the way on how to install git during installation, without running fork.exe?

The problem I am facing is that the application is installed in the `Program Files\Fork` folder which requires elevated permissions for unpacking **during first time run** & setting cred-manager.
Is there a way to force this setup during installation?

**EDIT:**
Seems like installation has been done manually, not with the `/s` flag.

Btw on fork.dev URL still points to 1.92, not 1.93

**Comment by jamaya77** [comment]
> @DanPristupov @jamaya77 Do you know the way on how to install git during installation, without running fork.exe?
> 
> The problem I am facing is that the application is installed in the `Program Files\Fork` folder which requires elevated permissions for unpacking **during first time run** & setting cred-manager. Is there a way to force this setup during installation?
> 
> **EDIT:** Seems like installation has been done manually, not with the `/s` flag.
> 
> Btw on fork.dev URL still points to 1.92, not 1.93


Hi Lonli-Lokli,
It's odd that your installation is pointing to " \Program Files\Fork". I just installed version 1.91 and the default source folder is still in "C:\Users\username". Are you using any specific parameters forcing to install as a system installation rather than the user?

Make sure when you're installing fork, that the deployment is running as the user.

As for "without using fork.exe", there is another way to deploy Fork without relying on a standalone installer. I use "winget" with a lot of my software deployment via sccm which will always fetch from online, download and install the latest stable version of any software.

[Fork](https://winget.run/pkg/Fork/Fork) is available from winget.


![Screenshot from 2024-01-22 17-27-42](https://github.com/fork-dev/TrackerWin/assets/80473349/be344f1f-c450-4853-abf0-e881e42a2edc)

**Comment by DanPristupov** [comment]
@Lonli-Lokli the installer must be run without elevated permissions, otherwise it will be installed to %programdata% and there will not be no way to proceed.

**Comment by Lonli-Lokli** [comment]
@DanPristupov The problem is that it's installable by other user, so even with /s other users will not be able to run it. 

So in enterprise, there is usually a dedicated process of deployment, which runs sometimes by other machines remotely or by other users.

Is there any way to install fork to the folder where all the users will have full access (obviously except for %programdata%)

**Comment by DanPristupov** [comment]
> Is there any way to install fork to the folder where all the users will have full access (obviously except for %programdata%)

No. Fork must be installed per user.

**Comment by Lonli-Lokli** [comment]
Thanks. Do you have plans to support all steps from there? https://github.com/fork-dev/TrackerWin/issues/57#issuecomment-832707759

In case of enterprise one more aspect is git certificates.
In order to support internal resources there is usually a company-internal ca-bundle.crt file, while Fork will use usual git and will not be able to clone internal repositories

**Comment by DanPristupov** [comment]
@Lonli-Lokli 

> Do you have plans to support all steps from there? https://github.com/fork-dev/TrackerWin/issues/57#issuecomment-832707759

System wide installing and ability to configure before install will not appear in the near future (this is a technical problem). The rest seems to be implemented already.

**Comment by Lonli-Lokli** [comment]
@DanPristupov So does it mean that 
1) this script below will run and Fork will NOT fail during unpacking git (what is happening now with the installation to the `Program Files`? I understand that auto-update will not work
2) Is it possible to put auto-update=off into `setup.config` too?
3) What other options are suppoted in this `setup.config`

```
@ECHO OFF
 
:: Run Fork installer NOT silently, point to the Program Files
ForkInstaller.exe 
 
:: Prepare the Fork settings file which will be used on the first run
(
echo "git=C:\program files\git\bin\git.exe"
) > %PROGRAMFILES%/Fork/setup.config

**Comment by DanPristupov** [comment]
@Lonli-Lokli as I said (https://github.com/fork-dev/TrackerWin/issues/57#issuecomment-1905774843), system wide installation (like to `Program Files`)  and pre-configuration (like setup.config) are NOT supported.

**Comment by Lonli-Lokli** [comment]
Sorry, my fault :) didn't read well. Ergh, it makes a problem for any user 8n a restricted env...

No way to skip git unpacking? Right now it's just do not allow to run Fork and select git instance manually

**Comment by Lonli-Lokli** [comment]
@DanPristupov initial configuration (especially part with disabling autoupdate and using system got) is really missing.

Can you use this workaround with squirrel?
https://github.com/Squirrel/Squirrel.Windows/issues/483#issuecomment-165483477

**Comment by 2PintChristianN** [comment]
Global install (per machine, and not per user) installs is a requirement for us to allow this in our environment.
This generally also require that Squirrel is not used, Squirrel is great for per user deployments, but here we need to be in control and update all versions on all machines. If things are installed per user it's not possible since we cant run as each user on each machine.
(this becomes a real issue when there is CVEs for libssl and have to eliminate all instances of the vuln dlls. using shared install of git would minimize those occurrences)

This also ties into that it needs to be installable, and uninstallable, in a reliable silent fashion, per machine (not user) thru Chocolatey.

@DanPristupov You keep saying this isn't supported, well for us this will be a requirement to allow users to buy the license.

**Comment by bleidd** [comment]
@DanPristupov I don’t think this will get anywhere near an enterprise environment without proper MSI/MSIX packaging.

Sure, 100 users might say “Priviet”, but that definitely won’t unlock any license budget 😉
